src/layers/domain/core/device_key/v2.py

```diff
@@ -14,4 +15,21 @@ class DeviceKey(BaseModel):
     @validator("key_value", check_fields=True)
     def validate_key(key_value: str, values: dict):
         key_type: DeviceKeyType = values.get("key_type")
-        return validate_key(key=key_value, type=key_type)
+        return validate_key(key_value=key_value, key_type=key_type)
+
+    @property
+    def parts(self):
+        return (self.key_type, self.key_value)
+
+    def __hash__(self):
+        return hash(self.parts)
+
+
+def validate_key(key_value: str, key_type: DeviceKeyType):
+    if key_type and key_type.pattern.match(key_value) is None:
+        raise InvalidDeviceKeyError(
+            f"Key '{key_value}' does not match the expected "
+            f"pattern '{key_type.pattern.pattern}' associated with "
```

Comment:
Not even a real issue but pattern.pattern looks a bit funny

Reply:
agreed, but it's an object of python type `re.Pattern` and it has a member called `pattern`